File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-6.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int res;
    private int dfs(TreeNode curr) {
        if (curr == null) {
            return 0;
        }

        int left = Math.max(dfs(curr.left), 0);
        int right = Math.max(dfs(curr.right), 0);

        res = Math.max(curr.val + left + right, res);

        return curr.val + Math.max(left, right);
    }

    public int maxPathSum(TreeNode root) {
        res = Integer.MIN_VALUE;

        dfs(root);

        return res;
    }
}

File: Data Structures & Algorithms/permutation-string/submission-6.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        unordered_map<char, int> s1Count{};

        for (char c: s1) {
            ++s1Count[c];
        }

        int left = 0;
        int right = 0;

        unordered_map<char, int> s2Count{};

        for (; right < s2.size(); ++right) {
            ++s2Count[s2[right]];

            for (;s2Count[s2[right]] > s1Count[s2[right]]; ++left) {
                --s2Count[s2[left]];
            }

            if (right - left == s1.size() - 1) {
                return true;
            }
        }
        
        return false;
    }
};
